src/main/java/com/williamfiset/algorithms/graphtheory/AStar_GridHeuristic.java

```diff
@@ -1,279 +1,229 @@
-    @Override
-    public int compareTo(Node other) {
-      if (Math.abs(f - other.f) < EPS) {
-        if (Math.abs(h - other.h) < EPS) return 0;
-        return (h - other.h) > 0 ? +1 : -1;
-      }
-      return (f - other.f) > 0 ? +1 : -1;
+    public static <A, B> Pair<A, B> of(A a, B b) {
+      return new Pair<>(a, b);
     }
   }
 
-  // Run A* algorithm on a directed graph to find the shortest path
-  // from a starting node to an ending node. If there is no path between the
-  // starting node and the destination node the returned value is set to be
-  // Double.POSITIVE_INFINITY.
-  public static double astar(
-      double[] X, double[] Y, Map<Integer, List<Edge>> graph, int start, int end, int n) {
-
-    // In the event that you wish to rebuild the shortest path
-    // you can do so using the prev array and starting at some node 'end'
-    // and finding the previous node using prev[end] and the previous node
-    // after that prev[prev[end]] etc... working all the way back until
-    // the index of start is found. Simply uncomment where the prev array is used.
-    // int[] prev = new int[n];
-
-    double[] G = new double[n];
-    double[] F = new double[n];
-    Arrays.fill(G, Double.POSITIVE_INFINITY);
-    Arrays.fill(F, Double.POSITIVE_INFINITY);
-    G[start] = 0;
-    F[start] = heuristic(X, Y, start, end);
-
-    Set<Integer> openSet = new HashSet<>();
-    Set<Integer> closedSet = new HashSet<>();
-
-    // Keep a priority queue of the next most promising node to visit
-    PriorityQueue<Node> pq = new PriorityQueue<>();
-    pq.offer(new Node(start, G[start], F[start]));
-    openSet.add(start);
-
-    while (!pq.isEmpty()) {
-
-      Node node = pq.poll();
-      openSet.remove(node.id);
-      closedSet.add(node.id);
-
-      if (node.id == end) return G[end];
-
-      List<Edge> edges = graph.get(node.id);
-      if (edges != null) {
-        for (int i = 0; i < edges.size(); i++) {
-
-          Edge edge = edges.get(i);
-          if (closedSet.contains(edge.to)) continue;
+  // Create a graph with V vertices
+  @SuppressWarnings("unchecked")
+  public static List<Edge>[] createGraph(final int V) {
+    List<Edge>[] graph = new List[V];
+    for (int i = 0; i < V; i++) graph[i] = new ArrayList<>();
+    return graph;
+  }
 
-          double g = node.g + edge.cost;
-          double h = heuristic(X, Y, edge.to, end);
+  // Helper function to add an edge to the graph
+  public static void addEdge(List<Edge>[] graph, int from, int to, int cost) {
+    graph[from].add(new Edge(to, cost));
+  }
 
-          if (g < G[edge.to] || !openSet.contains(edge.to)) {
+  /**
+   * The main A* algorithm.
+   *
+   * @param graph The graph containing the vertices and adjacency list.
+   * @param start The starting vertex.
+   * @param end The target vertex.
+   * @return The smallest distance with the weights from the start to end.
+   */
+  public static int aStar(List<Edge>[] graph, int start, int end) {
```

Comment:
I suggest returning a long instead of an int to avoid overflow as much as possible.